src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.Keyed.cs

```diff
@@ -0,0 +1,416 @@
+            ServiceCollectionDescriptorExtensions.TryAdd(collection, descriptor);
+        }
+
+        /// <summary>
+        /// Adds the specified <typeparamref name="TService"/> as a <see cref="ServiceLifetime.Scoped"/> service
+        /// to the <paramref name="collection"/> if the service type hasn't already been registered.
+        /// </summary>
+        /// <typeparam name="TService">The type of the service to add.</typeparam>
+        /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="serviceKey">The service key.</param>
+        public static void TryAddKeyedScoped<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TService>(this IServiceCollection collection, object? serviceKey)
+            where TService : class
+        {
+            ThrowHelper.ThrowIfNull(collection);
+
+            TryAddKeyedScoped(collection, typeof(TService), serviceKey, typeof(TService));
+        }
+
+        /// <summary>
+        /// Adds the specified <typeparamref name="TService"/> as a <see cref="ServiceLifetime.Scoped"/> service
+        /// implementation type specified in <typeparamref name="TImplementation"/>
+        /// to the <paramref name="collection"/> if the service type hasn't already been registered.
+        /// </summary>
+        /// <typeparam name="TService">The type of the service to add.</typeparam>
+        /// <typeparam name="TImplementation">The type of the implementation to use.</typeparam>
+        /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="serviceKey">The service key.</param>
+        public static void TryAddKeyedScoped<TService, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TImplementation>(this IServiceCollection collection, object? serviceKey)
+            where TService : class
+            where TImplementation : class, TService
+        {
+            ThrowHelper.ThrowIfNull(collection);
+
+            TryAddKeyedScoped(collection, typeof(TService), serviceKey, typeof(TImplementation));
+        }
+
+        /// <summary>
+        /// Adds the specified <typeparamref name="TService"/> as a <see cref="ServiceLifetime.Scoped"/> service
+        /// using the factory specified in <paramref name="implementationFactory"/>
+        /// to the <paramref name="services"/> if the service type hasn't already been registered.
+        /// </summary>
+        /// <typeparam name="TService">The type of the service to add.</typeparam>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="implementationFactory">The factory that creates the service.</param>
+        /// <param name="serviceKey">The service key.</param>
+        public static void TryAddKeyedScoped<TService>(
+            this IServiceCollection services,
+            object serviceKey,
```

Comment:
Should this be `object? serviceKey` for pass-through support (like elsewhere in this PR)?

Reply:
The `implementationFactory` signature is different if you are using a non-keyed registration, so I think it's better to not allow `null` in this case.